Use pre-commit in CI

chillenzer opened the pull request:
Implements most of the stuff suggested in #4793 and thereby closes #4793.

One very important hook I have left out: https://github.com/koalaman/shellcheck
This has a whole lot of great heuristics to write somewhat less fragile bash. But given the current discrepancy between its ideas of robust `bash` and the actual state of the repository it would need some automation to harmonise the two.

Also `ruff` has a lot to offer in terms of linting rules for Python. I haven't done any tinkering in that direction but there is a lot great advice in there.

chillenzer committed:
Use pre-commit in CI and remove some redundancy from check_cpp_code_style.sh

chillenzer committed:
De-bump version of pre-commit because Ubuntu is not up-to-date enough

chillenzer committed:
Test: Trigger CI failure with pre-commit

chillenzer committed:
Revert "Test: Trigger CI failure with pre-commit"

I leave this in to document that and how this was tested.

chillenzer committed:
Add pre-commit hook to remove tabs

chillenzer committed:
Add pre-commit hook to forbid and fix CRLF

Commit:
Remove CRLF line endings

chillenzer committed:
Remove redundant checks

chillenzer committed:
Manually remove non-ASCII

chillenzer committed:
Add minimum required pre-commit version

chillenzer committed:
Remove hasNonASCII

chillenzer committed:
Add check_cpp_code_style.sh to pre-commit

chillenzer committed:
Remove manually run check_cpp_code_style.sh

chillenzer committed:
Test: Trigger CI failure for Doxygen

chillenzer committed:
Revert "Test: Trigger CI failure for Doxygen"

As before, I leave the previous in to document that and how this was
tested.

chillenzer committed:
Update docs about pre-commit

chillenzer committed:
Introduce ruff for python linting

Commit:
Run ruff format

chillenzer committed:
Move imports to top

chillenzer committed:
Bump pre-commit hooks version

chillenzer committed:
Make python scripts executable

chillenzer committed:
Add more hooks

chillenzer committed:
Add executables have shebang hook

chillenzer committed:
Fix file permissions (remove executable)

chillenzer committed:
Add shebangs

chillenzer committed:
Fix more file permissions (remove executable)

PrometheusPi commented:
@chillenzer there are still some CI issues.

chillenzer commented:
I'm aware. It's a minor change in a Python script where I change `type(...) == ...` to `isinstance(...)` while following a linter's suggestion. Apparently, I forgot a `not` or something. Gotta check in detail.

chillenzer committed:
Ignore E721 in pre-commit

chillenzer committed:
Fix typo

chillenzer commented:
The refactoring issue is fixed (more precisely: reverted). One further note: If files are touched as a whole, this most likely comes down to the fact that I fixed permissions of some files.

chillenzer commented:
Hi, CI failed due to [connection issues](https://gitlab.com/hzdr/crp/picongpu/-/jobs/6214387641) (?) or the like. Please restart CI.

PrometheusPi commented:
@chillenzer You changed /added 86 files, some seem to have no connection to the pre-commit for the CI, like this [atomic physics image](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4817/files#diff-ec588138b77b1e5a7c61e188a3748ab292d8d0eb17e7ff4f7a17824eb6c6d6fe)
Please have a look at your pull request and clean up the extra files

psychocoderHPC commented:
> @chillenzer You changed /added 86 files, some seem to have no connection to the pre-commit for the CI, like this [atomic physics image](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4817/files#diff-ec588138b77b1e5a7c61e188a3748ab292d8d0eb17e7ff4f7a17824eb6c6d6fe) Please have a look at your pull request and clean up the extra files

Most likely he fixed the access rights of the file. @chillenzer Is this correct?

chillenzer commented:
Yes, as already stated here:

> The refactoring issue is fixed (more precisely: reverted). One further note: If files are touched as a whole, this most likely comes down to the fact that I fixed permissions of some files.

PrometheusPi commented on lib/python/picongpu/extra/plugins/jupyter_widgets/base_widget.py:
```diff
@@ -359,8 +359,9 @@ def _update_available_sim_times(self):
             print(
-                "last selected time {0} not present with newly"
-                " selected sim_times, set to {1}".format(last_selected_val, common_sim_times[0])
+                "last selected time {0} not present with newly" " selected sim_times, set to {1}".format(
```
This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

PrometheusPi commented on lib/python/picongpu/extra/plugins/data/transitionradiation.py:
```diff
@@ -257,8 +258,9 @@ def get_data(self, iteration=None, **kwargs):
             print(
-                "Angular intensity distribution is sliced at theta={:.2e} "
-                "with omega={:.2e}.".format(self.thetas[theta], self.omegas[omega])
+                "Angular intensity distribution is sliced at theta={:.2e} " "with omega={:.2e}.".format(
```
This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

PrometheusPi commented on lib/python/picongpu/extra/plugins/data/transitionradiation.py:
```diff
@@ -239,8 +239,9 @@ def get_data(self, iteration=None, **kwargs):
             print(
-                "Angular intensity distribution is sliced at phi={:.2e} "
-                "with omega={:.2e}.".format(self.phis[phi], self.omegas[omega])
+                "Angular intensity distribution is sliced at phi={:.2e} " "with omega={:.2e}.".format(
```
This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

PrometheusPi commented on lib/python/picongpu/extra/plugins/data/png.py:
```diff
@@ -192,9 +192,9 @@ def _get_for_iteration(self, iteration, species, species_filter="all", axis=None
-                    "Iteration {} is not available!\n"
-                    "List of available iterations: \n"
-                    "{}".format(iteration, available_iterations)
+                    "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

PrometheusPi commented on lib/python/picongpu/extra/plugins/data/phase_space.py:
```diff
@@ -198,9 +198,9 @@ def _get_for_iteration(self, iteration, ps, species, species_filter="all", file_
-                    "Iteration {} is not available!\n"
-                    "List of available iterations: \n"
-                    "{}".format(iteration, available_iterations)
+                    "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

PrometheusPi commented on lib/python/picongpu/extra/plugins/data/energy_histogram.py:
```diff
@@ -159,9 +159,9 @@ def _get_for_iteration(self, iteration, species, species_filter="all", include_o
-                "Iteration {} is not available!\n"
-                "List of available iterations: \n"
-                "{}".format(iteration, data.index.values)
+                "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
here too

PrometheusPi commented on lib/python/picongpu/extra/plugins/data/emittance.py:
```diff
@@ -163,9 +163,9 @@ def _get_for_iteration(self, iteration, species, species_filter="all", **kwargs)
-                "Iteration {} is not available!\n"
-                "List of available iterations: \n"
-                "{}".format(iteration, data.index.values)
+                "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
here too

PrometheusPi commented on lib/python/picongpu/extra/input/parameters.py:
```diff
@@ -145,16 +146,18 @@ def _check_input(self, vals):
                 raise ValueError(
-                    "Invalid values found! Values should be "
-                    "contained in {0} but are {1}!".format(self.pic_range, vals)
+                    "Invalid values found! Values should be " "contained in {0} but are {1}!".format(
```
here too

PrometheusPi commented on lib/python/picongpu/extra/input/parameters.py:
```diff
@@ -145,16 +146,18 @@ def _check_input(self, vals):
                 raise ValueError(
-                    "Invalid values found! Values should be elements of "
-                    "{0} but are {1}!".format(self.pic_values, vals)
+                    "Invalid values found! Values should be elements of " "{0} but are {1}!".format(
```
here too

PrometheusPi commented on include/picongpu/plugins/openPMD/GetComponentsType.hpp:
```diff
@@ -1,50 +1,50 @@
-
-    } // namespace traits
-} // namespace pmacc
+/* Copyright 2021-2023 Sergei Bastrakov
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

PrometheusPi commented on share/picongpu/examples/LaserWakefield/include/picongpu/param/incidentField.param:
```diff
@@ -1,342 +1,342 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov, Rene Widera
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

chillenzer replied:
`git blame` reveals that the change stems from f29c574a5 which removed CRLF line endings (as the commit message quite literally states). That's why each and every line is different. **They all have to end...** (That sounded way too ominous to _not _ end with it.)

PrometheusPi commented on share/picongpu/examples/FoilLCT/include/picongpu/param/incidentField.param:
```diff
@@ -1,331 +1,331 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov, Rene Widera
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

PrometheusPi commented on share/picongpu/examples/Bunch/include/picongpu/param/incidentField.param:
```diff
@@ -1,229 +1,229 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov, Rene Widera
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

PrometheusPi commented on share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param:
```diff
@@ -1,603 +1,603 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2014-2023 Axel Huebl, Alexander Debus, Richard Pausch, Sergei Bastrakov
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/operation/operation.py:
```diff
@@ -180,8 +180,9 @@ def bake_species_attributes(self) -> None:
                 raise ValueError(
-                    "only on attribute per type is allowed per "
-                    "species, offending: {}".format(", ".join(duplicate_names))
+                    "only on attribute per type is allowed per " "species, offending: {}".format(
```
same here

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/operation/momentum/drift.py:
```diff
@@ -45,8 +45,9 @@ def __check_vector_real(self, vector: typing.Tuple[float, float, float]) -> None
                 raise ValueError(
-                    "vector may only contain real components, "
-                    "offending axis: {}".format(["x", "y", "z"][vector.index(invalid)])
+                    "vector may only contain real components, " "offending axis: {}".format(
```
same here

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/initmanager.py:
```diff
@@ -237,8 +238,7 @@ def __check_species_dependencies_circular(self) -> None:
                 raise RecursionError(
-                    "species {} is in circular dependency, "
-                    "all dependencies are: {}".format(
+                    "species {} is in circular dependency, " "all dependencies are: {}".format(
```
same here

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/initmanager.py:
```diff
@@ -336,10 +336,10 @@ def __check_constant_attribute_dependencies(self) -> None:
-                    ), "constant {} of species {} requires attribute {} to " "be present, but it is not".format(
-                        constant, species.name, required_attr
+                    assert required_attr.PICONGPU_NAME in species_attr_names, (
+                        "constant {} of species {} requires attribute {} to " "be present, but it is not".format(
```
same here

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/initmanager.py:
```diff
@@ -372,10 +372,10 @@ def __check_constant_constant_dependencies(self):
-                    ), "species {}: required constant {} not found, " "(required by constant {})".format(
-                        species.name, required_constant, constant
+                    assert species.has_constant_of_type(required_constant), (
+                        "species {}: required constant {} not found, " "(required by constant {})".format(
```
same here

PrometheusPi commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -497,8 +496,9 @@ def write_input_file(self, file_name: str) -> None:
             raise ValueError(
-                "PIConGPU does not support stepwise running. Invoke step() "
-                "with max_steps (={})".format(self.max_steps)
+                "PIConGPU does not support stepwise running. Invoke step() " "with max_steps (={})".format(
```
same here

PrometheusPi commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -375,8 +373,9 @@ def __get_electron_species(self) -> PicongpuPicmiSpecies:
             logging.info(
-                "no electron species for ionization available, "
-                "creating electrons with name: {}".format(electrons_name)
+                "no electron species for ionization available, " "creating electrons with name: {}".format(
```
same here

PrometheusPi commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -99,7 +97,7 @@ def __init__(
         picongpu_template_dir: typing.Optional[typing.Union[str, pathlib.Path]] = None,
         picongpu_typical_ppc: typing.Optional[int] = None,
-        **kw
+        **kw,
```
not essential: why is a comma after the last argument preferred?

PrometheusPi commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -70,9 +70,7 @@ def __yee_compute_cfl_or_delta_t(self) -> None:
-            delta_t_from_cfl = self.solver.cfl / (
-                constants.c * sqrt(1 / delta_x**2 + 1 / delta_y**2 + 1 / delta_z**2)
-            )
+            delta_t_from_cfl = self.solver.cfl / (constants.c * sqrt(1 / delta_x**2 + 1 / delta_y**2 + 1 / delta_z**2))
```
same here

PrometheusPi commented on lib/python/picongpu/picmi/gaussian_laser.py:
```diff
@@ -58,7 +56,7 @@ def __init__(
             [16, -16],
         ],
-        **kw
+        **kw,
```
strange comma

chillenzer replied:
It's a convention used by `black` as well as `ruff format` (which this PR is attempting to switch to). The reasoning for this is concisely shown by example [here](https://stackoverflow.com/questions/12087742/should-i-add-a-trailing-comma-after-the-last-argument-in-a-function-call/17492103#17492103).

PrometheusPi commented on lib/python/picongpu/picmi/gaussian_laser.py:
```diff
@@ -86,7 +84,7 @@ def __init__(
             focal_position,
             centroid_position,
-            **kw
+            **kw,
```
another strange comma

PrometheusPi commented on include/picongpu/param/incidentField.param:
```diff
@@ -1,180 +1,180 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on include/picongpu/fields/incidentField/profiles/profiles.hpp:
```diff
@@ -1,29 +1,29 @@
-#include "picongpu/fields/incidentField/profiles/PlaneWave.hpp"
-#include "picongpu/fields/incidentField/profiles/Polynom.hpp"
-#include "picongpu/fields/incidentField/profiles/Wavepacket.hpp"
+/* Copyright 2022-2023 Sergei Bastrakov
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on include/picongpu/fields/incidentField/profiles/profiles.def:
```diff
@@ -1,31 +1,31 @@
-#include "picongpu/fields/incidentField/profiles/PulseFrontTilt.def"
-#include "picongpu/fields/incidentField/profiles/Traits.hpp"
-#include "picongpu/fields/incidentField/profiles/Wavepacket.def"
+/* Copyright 2022-2023 Sergei Bastrakov
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on include/picongpu/fields/incidentField/profiles/Wavepacket.hpp:
```diff
@@ -1,185 +1,185 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Axel Huebl, Heiko Burau, Rene Widera, Richard Pausch,
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on include/picongpu/fields/incidentField/profiles/Wavepacket.def:
```diff
@@ -1,87 +1,87 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Axel Huebl, Heiko Burau, Rene Widera, Richard Pausch,
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on include/picongpu/fields/incidentField/profiles/Traits.hpp:
```diff
@@ -1,44 +1,44 @@
-        = MakePeriodicTransversalHuygensSurfaceContiguous<T_Profile>::value;
-
-} // namespace picongpu::fields::incidentField::profiles
+/* Copyright 2023-2023 Rene Widera
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on include/picongpu/fields/incidentField/profiles/PulseFrontTilt.def:
```diff
@@ -1,69 +1,69 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Anton Helm, Heiko Burau, Rene Widera, Richard Pausch,
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on include/picongpu/fields/incidentField/profiles/Polynom.hpp:
```diff
@@ -1,170 +1,170 @@
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Heiko Burau, Rene Widera, Richard Pausch, Axel Huebl, Sergei Bastrakov
```
This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

PrometheusPi commented on .pre-commit-config.yaml:
```diff
@@ -7,23 +8,64 @@ repos:
+    # Run the formatter.
+    - id: ruff-format
+      types_or: [ python, pyi, jupyter ]
+      args: ["--line-length", "120"]
```
This is a change from our previous 79 characters per line (PEP8) limit

chillenzer replied:
Yes, this was agreed with @psychocoderHPC and I thought that it was already agreed upon elsewhere. But I've just tried to look it up but admittedly I could neither find it in `CONTRIBUTING.md` nor in our [Contributing repo](https://github.com/ComputationalRadiationPhysics/contributing). No strong feelings here. Actually, I would personally not change it.

PrometheusPi commented on docs/COMMIT.md:
```diff
@@ -33,7 +33,9 @@ $ pre-commit install
+- You can run all hooks on all files via `pre-commit run --all-files [--hook-stage manual]`. The last argument
+  `--hook-stage manual` includes rather slow additional tests that are run by the CI but are considered too heavy-weight
+  to run before each commit.
 - If a check fails, oftentimes it will automatically apply the fixes (e.g. run `clang-format`). If there are no unstaged
```
optional (not part of this PR): 
```suggestion
- If a check fails, the fixes are often applied automatically (e.g., run `clang-format`). If there are no unstaged
```

PrometheusPi reviewed (changes requested):
looks god to me (except that GitHub does not show right changes correctly all the times). 
Only the characters per line increase from 79 to 120 causes some wired code style that should be fixed (or we go back to 79 characters)

chillenzer committed:
Adjust strings to new line length

chillenzer committed:
Rephrasing in pre-commit docs

Co-authored-by: Richard Pausch <r.pausch@hzdr.de>
